Add graceful behavior when processes are killed

naddeoa opened the pull request:
When working with gunicorn I discovered that you'll end up shooting yourself in the foot if you start the process logger at the wrong time because gunicorn will end up killing the logger, which results in apparently nothing happening after you make log calls. This will update the logger to be explicit about how dead it is and why, throwing on log() and close() calls.

Commit:
Add graceful behavior when processes are killed

When working with gunicorn I discovered that you'll end up shooting
yourself in the foot if you start the process logger at the wrong time
because gunicorn will end up killing the logger, which results in
apparently nothing happening after you make log calls. This will update
the logger to be explicit about how dead it is and why, throwing on
log() and close() calls.

jamie256 reviewed (commented):
The tests seem to be taking a long time and not progressing, can you look into the CI issues?

https://github.com/whylabs/whylogs/actions/runs/6713659241/job/18267898666

Commit:
Update sigint ignoring in rolling logger process

Apparently it wasn't enough to just block signals with pthread_sigmask
because ctrl-c would still pass through and end up killing the profiling
process. You also have to set up signal handlers that ignore those things.

Commit:
Add ability to customize logger creation logic in process logger

This lets you customize how the thread rolling loggers are created in
the process rolling logger, in case you need different upload/aggregate
configs for each dataset id.

naddeoa commented:
@jamie256 Finally got this working again with a few additional process logger changes needed for the container.

jamie256 reviewed (approved):
LGTM!
